src/main/scala/mesosphere/marathon/state/GroupManager.scala

```diff
@@ -218,19 +219,22 @@ class GroupManager @Inject() (
       port
     }
 
-    def mergeServicePortsAndPortDefinitions(portDefinitions: Seq[PortDefinition], servicePorts: Seq[Int]) = {
-      portDefinitions.zipAll(servicePorts, AppDefinition.RandomPortDefinition, AppDefinition.RandomPortValue).map {
-        case (portDefinition, servicePort) => portDefinition.copy(port = servicePort)
-      }
-    }
+    def mergeServicePortsAndPortDefinitions(
+      portDefinitions: Seq[PortDefinition],
+      servicePorts: Seq[Int]): Seq[PortDefinition] =
+      if (portDefinitions.nonEmpty)
+        portDefinitions.zipAll(servicePorts, AppDefinition.RandomPortDefinition, AppDefinition.RandomPortValue).map {
+          case (portDefinition, servicePort) => portDefinition.copy(port = servicePort)
+        }
+      else Seq.empty
 
     def assignPorts(app: AppDefinition): AppDefinition = {
 
       //all ports that are already assigned in old app definition, but not used in the new definition
       //if the app uses dynamic ports (0), it will get always the same ports assigned
       val assignedAndAvailable = mutable.Queue(
         from.app(app.id)
-          .map(_.portNumbers.filter(p => portRange.contains(p) && !app.servicePorts.contains(p)))
+          .map(_.hostPorts.flatten.filter(p => portRange.contains(p) && !app.servicePorts.contains(p)))
```

Comment:
that depends: how wrong was the original code? substituting `hostPorts` for `portNumbers` uses host ports from portmappings when there's a container. using `servicePorts` instead sounds like a different semantic